test/e2e_node/image_pull_test.go

```diff
@@ -0,0 +1,241 @@
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeletconfig "k8s.io/kubernetes/pkg/kubelet/apis/config"
+	kubeletevents "k8s.io/kubernetes/pkg/kubelet/events"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	"k8s.io/kubernetes/test/e2e/nodefeature"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	admissionapi "k8s.io/pod-security-admission/api"
+	"k8s.io/utils/ptr"
+)
+
+// This test needs to run in serial to prevent caching of the images by other tests
+// and to prevent the wait time of image pulls to be increased by other images
+var _ = SIGDescribe("Pull Image", framework.WithSerial(), nodefeature.MaxParallelImagePull, func() {
+
+	f := framework.NewDefaultFramework("parallel-pull-image-test")
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+	var testpods []*v1.Pod
+
+	ginkgo.Context("parallel image pull with MaxParallelImagePulls=5", func() {
+		tempSetCurrentKubeletConfig(f, func(ctx context.Context, initialConfig *kubeletconfig.KubeletConfiguration) {
+			initialConfig.SerializeImagePulls = false
+			initialConfig.MaxParallelImagePulls = ptr.To[int32](5)
+		})
+
+		ginkgo.BeforeEach(func(ctx context.Context) {
+			testpods = prepareAndCleanup(ctx, f)
+		})
+
+		ginkgo.AfterEach(func(ctx context.Context) {
+			ginkgo.By("cleanup pods")
+			for _, pod := range testpods {
+				deletePodSyncByName(ctx, f, pod.Name)
+			}
+		})
+
+		ginkgo.It("should pull immediately if no more than 5 pods", func(ctx context.Context) {
+			for _, testpod := range testpods {
+				pod := e2epod.NewPodClient(f).Create(ctx, testpod)
+				err := e2epod.WaitForPodCondition(ctx, f.ClientSet, f.Namespace.Name, pod.Name, "Running", 30*time.Second, func(pod *v1.Pod) (bool, error) {
+					if pod.Status.Phase == v1.PodRunning {
+						return true, nil
+					}
+					return false, nil
+				})
+				framework.ExpectNoError(err)
+			}
+
+			events, err := f.ClientSet.CoreV1().Events(f.Namespace.Name).List(ctx, metav1.ListOptions{})
+			framework.ExpectNoError(err)
+			var httpdPulled []*pulledStruct
+			for _, event := range events.Items {
+				if event.Reason == kubeletevents.PulledImage {
+					for _, testpod := range testpods {
+						if event.InvolvedObject.Name == testpod.Name {
+							pulled, err := getDurationsFromPulledEventMsg(event.Message)
+							httpdPulled = append(httpdPulled, pulled)
+							framework.ExpectNoError(err)
+							break
+						}
+					}
+				}
+			}
+			gomega.Expect(len(testpods)).To(gomega.BeComparableTo(len(httpdPulled)))
+
+			// as this is parallel image pulling, the waiting duration should be similar with the pulled duration.
+			// use 1.2 as a common ratio
+			for _, pulled := range httpdPulled {
+				if float32(pulled.pulledIncludeWaitingDuration/time.Millisecond)/float32(pulled.pulledDuration/time.Millisecond) > 1.2 {
```

Comment:
and similarly this should be "< 1.2"?

Reply:
> and similarly this should be "< 1.2"?

The expected similarly should be "<= 1.2", and so we fail if it is`>1.2`